src/libmongoc/src/mongoc/uthash.h

```diff
@@ -0,0 +1,1145 @@
+/*
+Copyright (c) 2003-2021, Troy D. Hanson     http://troydhanson.github.com/uthash/
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
+OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#ifndef UTHASH_H
+#define UTHASH_H
+
+// libmongoc specific code ... begin
+// clang-format off
+#include <mongoc-prelude.h>
+#include <bson/bson.h>
+// Use libbson's malloc/free functions in uthash.
+#define uthash_malloc(sz) bson_malloc (sz)
+#define uthash_free(ptr, sz) bson_free (ptr)
+// libmongoc specific code ... end
```

Comment:
Rather than modifying bundled source files, consider using an indirect header file instead:
```c
// src/libmongoc/src/mongoc/uthash.h

#ifndef MONGOC_UTHASH_H
#define MONGOC_UTHASH_H

#include <mongoc-prelude.h>
#include <bson/bson.h>

#define uthash_malloc(sz) bson_malloc (sz)
#define uthash_free(ptr, sz) bson_free (ptr)

#include <uthash-2.3.0/uthash.h>

#endif // MONGOC_UTHASH_H
```

Similarly for the existing `utlist.h` header.

Reply:
Thank you for the suggestion. I very much like the idea of an indirect header. That enables keeping an unmodified `uthash.h` header. Suggestion has been applied for both `uthash.h` and `utlist.h`.